core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -690,12 +693,18 @@ public PlanWithProperties visitMergeWriter(MergeWriterNode node, PreferredProper
 
         private PlanWithProperties getWriterPlanWithProperties(Optional<PartitioningScheme> partitioningScheme, PlanWithProperties newSource, TableWriterNode.WriterTarget writerTarget)
         {
+            // use maxWritersTasks to set PartitioningScheme.partitionCount field to limit number of tasks that will take part in executing writing stage
+            // TODO: Add support for FTE
+            Optional<Integer> maxWritersNodesCount = getRetryPolicy(session) != RetryPolicy.TASK
+                    ? Optional.of(writerTarget.getMaxWriterTasks(plannerContext.getMetadata(), session).orElse(getMaxWriterTasksCount(session)))
```

Comment:
It is how `PlanFragmenter` works. In `LogicalPlanner`. To set partitioning of fragment, `PlanFragmenter` traverse plan top -> bottom and "records" partitioning from nodes. Because exchanges are children in relate to `TableWriterNode`, this one decides about `partitioning` - because it is visited before exchanges are visisted. 

As I said at the beginning, it is just how `PlanFragmenter` works.